### PR TITLE
Use lists to model paths and queries.

### PR DESCRIPTION
**Pull request by swankjesse**
This is actually a little trickier than strings, partly because
we now need code to go in both directions, and partly because
paths that end with '/' are an important but awkward special
case.

The goal is that this will make accessing query parameters much
easier.

### CODE REVIEW

**Comment by swankjesse** [comment]
https://github.com/square/okhttp/issues/1486

**Comment by JakeWharton** [comment]
Needs a rebase.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 1]
```diff
@@ -31,6 +31,8 @@
 /**
  * A <a href="https://url.spec.whatwg.org/">URL</a> with an {@code http} or {@code https} scheme.
  *
+ * TODO: parts of a URL
```
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Probably your conflict right here.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 2]
```diff
@@ -245,7 +320,19 @@ public HttpUrl resolve(String link) {
...
 
   public Builder newBuilder() {
-    return new Builder(this);
+    Builder result = new Builder();
+    result.scheme = scheme;
+    result.username = username;
+    result.password = password;
+    result.host = host;
+    result.port = port;
+    result.pathSegments.clear();
```
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Redundant?

Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Nope, needed 'cause it defaults to `""`.

Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Ah, missed the ctor line.

**Comment by swankjesse** [comment]
Rebased!

**Comment by JakeWharton** [comment]
LGTM :shipit:

**Pull request merged**